libndt7/src/main/java/net/measurementlab/ndt7/android/Uploader.kt

```diff
@@ -0,0 +1,116 @@
+import net.measurementlab.ndt7.android.utils.DataConverter.currentTimeInMicroseconds
+import net.measurementlab.ndt7.android.utils.DataConverter.generateResponse
+import net.measurementlab.ndt7.android.utils.NDT7Constants.MAX_RUN_TIME
+import net.measurementlab.ndt7.android.utils.NDT7Constants.MEASUREMENT_INTERVAL
+import net.measurementlab.ndt7.android.utils.NDT7Constants.MIN_MESSAGE_SIZE
+import net.measurementlab.ndt7.android.NDTTest.TestType.*
+import net.measurementlab.ndt7.android.models.CallbackRegistry
+import net.measurementlab.ndt7.android.models.Measurement
+import net.measurementlab.ndt7.android.utils.PayloadTransformer
+import net.measurementlab.ndt7.android.utils.SocketFactory
+import okhttp3.OkHttpClient
+import okhttp3.WebSocket
+import okhttp3.WebSocketListener
+import okio.ByteString
+import java.net.URI
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.Semaphore
+
+class Uploader(
+        private val cbRegistry: CallbackRegistry,
+        private val executorService: ExecutorService,
+        private val speedtestLock: Semaphore
+) : WebSocketListener() {
+
+    private var startTime: Long = 0
+    private var previous: Long = 0
+    private var totalBytesSent = 0.0
+    private val gson = Gson()
+
+    override fun onMessage(webSocket: WebSocket, text: String) {
+
+        try {
+            val measurement = gson.fromJson(text, Measurement::class.java)
+            cbRegistry.measurementProgressCbk(measurement)
+        } catch (e: Exception) {
+            //we don't care that much if a single measurement has trouble
+            return
+        }
+    }
+
+
+override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
+
+    val clientResponse = generateResponse(startTime, totalBytesSent, UPLOAD)
+    when (code) {
+        1000 -> {
+            cbRegistry.onFinishedCbk(clientResponse, null, UPLOAD)
+        }
+        else -> {
+            cbRegistry.onFinishedCbk(clientResponse, Error(reason), UPLOAD)
+        }
+    }
+
+    releaseResources()
+    webSocket.close(1000, null)
+}
+
+    override fun onFailure(webSocket: WebSocket, t: Throwable, response: okhttp3.Response?) {
+        cbRegistry.onFinishedCbk(generateResponse(startTime, totalBytesSent, UPLOAD), t, UPLOAD)
+
+        releaseResources()
+        webSocket.close(1001, null)
+    }
+
+    fun beginUpload(url: String, httpClient: OkHttpClient?) {
+        val ws: WebSocket = SocketFactory.establishSocketConnection(url, httpClient, this)
+        startTime = currentTimeInMicroseconds()
+        previous = startTime
+
+        createBytePayloads(ws)
+    }
+
+    private fun createBytePayloads(ws: WebSocket) {
+
+        val timerStart = currentTimeInMicroseconds()
+        var elapsedTime = currentTimeInMicroseconds() - timerStart
+
+        //only allow this to run for 10 seconds, then stop
+        while (elapsedTime < MAX_RUN_TIME) {
+            val bytes = ByteString.of(*ByteArray(MIN_MESSAGE_SIZE))/* (1<<13) */
```

Comment:
@ahardewig the `performDynamicTuning` implementation is great! My apologies for missing it from my previous review earlier! Now I see how this allocation of bytes here had confused me before.

Even now that I see there's dynamic tuning, I am confused by this allocation. Did you perhaps mean to put this allocation outside of the loop? If the allocation inside the loop is intended, then I have this comment:

The ideal behaviour is to monotonically scale the message while the upload is progressing. However, by recreating a `1<<13` message here at the top of the loop, it seems to me you're resetting the message size to `1<<13` every time you return from `sendToWebSocket`.

Technically the spec allows this way of scaling messages. But we determined long time ago with @pboothe that it was most likely more proper to never scale _down_ the message size because TCP is also scaling down its send rate when there are network issues. And we don't want these two reactive loops to interfere with one another, because this may cause performance issues (i.e. we end up sending too little).

This is why the spec includes and algorithm that increases monotonically.

Would it be possible to allocate the `1<<13` message inside the loop and then increase monotonically?

On our end, I think we should improve the wording in the spec, to clarify that scaling down the message is something that client SHOULD NOT (MUST NOT?) do (cc: @stephen-soltesz)

Reply:
Thanks for checking that out! Yeah, you are totally right. I think you found a bug. The allocation should happen outside of the loop or else the message size will go back to starting off at the MIN_MESSAGE_SIZE. I've pushed a commit just now that seems to resolve this. Let me know if this looks any better. I can see the upload size consistently increasing now to better saturate my upload